components/brave_wallet/resources/solana_provider.js

```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+(function () {
+  if (!window.solana) {
+    return
+  }
+  const solanaWeb3 = require('@solana/web3.js')
+  window.solana.createPublickey = function createPublickey(base58Str) {
+    const result = new Object()
+    result.publicKey = new solanaWeb3.PublicKey(base58Str)
```

Comment:
Looks like the dependency isn't validating the public key is on the curve in the constructor call of PublicKey here, but has an implementation to optionally do it. This could have subtle impacts on downstream usage of the publicKey later or alternatively be used in odd ways so the safest thing for us to do would be to validate this key is on the curve or error on the input supplied.

I'm thinking that we should add the call to check that the publicKey is valid to prevent a GIGO scenario from this API, but there would be a tradeoff in performance when calling this API since we'd need to run [these math ops](https://github.com/solana-labs/solana-web3.js/blob/70bb763a10e193fe97ef439593d60165c5a6eb80/src/publickey.ts#L218) each time the call is ran.

How performant do we need this call to be? 

Ideally, we'd validate the public key on input using `is_on_curve(result.publicKey)` and return the result if that returns true (they use 1 which JS should cast to true) otherwise throw an error and reject the input. If we need to keep this extremely performant then we may be able to get away with ignoring the valid curve check, but it's a bit of a code smell in my eyes.

Reply:
Input of this function is provided by us from browser process, that doesn't need any validation

Reply:
we used it like this way
https://github.com/brave/brave-core/pull/12554/files#diff-7cdd1da2bab576a68d9a1469b45add340ad978aafbdff369f3e590b1aaf8febbR210-R222
to wrap browser process result and return it to caller

Reply:
Ok, I'm not familiar enough with that part of the codebase yet, so didn't spot that part of it. Thank you for pointing that out to me. Based on what I understand then this code won't get exposed through the window.solana API and will only be accessible to the browser process I don't think this would be a problem any longer.

Reply:
It's actually our renderer process using that to wrap result from browser process. When our renderer process uses it, we have to call it like `window.solana.createPublickey` but in v8, that is the only way we can get solanaWeb3 object in renderer code. So this will be accessible to anyone who has access to `window.solana` unfortunately, maybe we can rename it as `_createPublicKey` and `_createTransaction` to indicate that is not intended for public use and gives a warning when it gets called. Even someone purposefully use it with garbage input and get garbage out or error that is not our problem, that internal API is just a bridge to solanaWeb3 object creation

Reply:
Ok this part of the code is starting to make a bit more sense to me now. After spending some time looking at the WebUI documents a bit more it makes sense why it would get exposed, but also why it's not really intended to be used by others.

>Even someone purposefully use it with garbage input and get garbage out or error that is not our problem, that internal API is just a bridge to solanaWeb3 object creation

This is actually a good point as well and after looking at the program derived addresses work yesterday where some of the programs publicKey's are off curve intentionally, it may actually be overly limiting to force that check always. This of course would depend on if this API gets used as an intermediary step for creating those addresses as well which I'm not entirely certain off, but that's a common pattern I found in Solana where this could be considered a feature and not a bug.

> maybe we can rename it as _createPublicKey and _createTransaction to indicate that is not intended for public use and gives a warning

These seem like a reasonable way to handle this. Now that I'm understanding some of the tradeoffs made here better I'd be alright with leaving the API calls without the `_` and just the warning if we need to strictly adhere to the phantom API design for interop reasons.

Reply:
thanks. warning message added in 787e745a7c285f94d9d40b80a599d4ce9ba33641